src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardDictionaryRemoveByContainsKey.cs

```diff
@@ -0,0 +1,155 @@
+            isDataflowRule: false,
+            additionalCustomTags: WellKnownDiagnosticTags.Unnecessary);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!TryGetDictionaryTypeAndContainsKeyMethod(context.Compilation, out var dictionaryType, out var containsKeyMethod))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context => AnalyzeOperation(context, dictionaryType, containsKeyMethod), OperationKind.Conditional);
+
+            static void AnalyzeOperation(OperationAnalysisContext context, INamedTypeSymbol dictionaryType, IMethodSymbol containsKeyMethod)
+            {
+                var conditionalOperation = (IConditionalOperation)context.Operation;
+
+                IInvocationOperation? invocationOperation = null;
+
+                switch (conditionalOperation.Condition)
+                {
+                    case IInvocationOperation iOperation:
+                        invocationOperation = iOperation;
+                        break;
+                    case IUnaryOperation unaryOperation when unaryOperation.OperatorKind == UnaryOperatorKind.Not:
+                        if (unaryOperation.Operand is IInvocationOperation operand)
+                            invocationOperation = operand;
+                        break;
+                    default:
+                        return;
+                }
+
+                if (invocationOperation == null || !invocationOperation.TargetMethod.OriginalDefinition.Equals(containsKeyMethod, SymbolEqualityComparer.Default))
+                {
+                    return;
+                }
+
+                if (conditionalOperation.WhenTrue.Children.Any())
+                {
+                    var properties = ImmutableDictionary.CreateBuilder<string, string?>();
+                    properties[ConditionalOperation] = CreateLocationInfo(conditionalOperation.Syntax);
+
+                    switch (conditionalOperation.WhenTrue.Children.First())
+                    {
+                        case IInvocationOperation childInvocationOperation:
+                            if (childInvocationOperation.TargetMethod.Name == "Remove" &&
+                                childInvocationOperation.TargetMethod.OriginalDefinition.ContainingType.Equals(dictionaryType, SymbolEqualityComparer.Default))
+                            {
+                                properties[ChildStatementOperation] = CreateLocationInfo(childInvocationOperation.Syntax.Parent);
```

Comment:
This feels like a very twisted way of doing things in the analyzer. Analyzer is going out of its way to encode couple of additional locations that it itself doesn't really need, but are needed just by the fixer. That doesn't seem right. Yes, if the analyzer itself had some logic where it needed these locations for analysis, and the fixer also needed the same locations, it would make sense to encode them in the diagnostic's properties. IMO, we don't need this encoding logic and instead can let the fixer just find the first parenting IConditionalOperation for the IInvocationOperation on which the diagnostic is reported.

Reply:
I also had [similar comment](https://github.com/dotnet/roslyn-analyzers/pull/4836#discussion_r584975970) previously, though i was OK with this solution because even if it can be populated in the fixer, this might be easier solution as there is 2 locations needs an update, also @chucker pointed out an existing analyzer https://github.com/dotnet/roslyn-analyzers/blob/0c025fe9790856b943fe341243e92e680b6a0ac4/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs#L549 that using this approach, so i assumed this approach is acceptable

Reply:
@buyaa-n That implementation is done to work around a Roslyn bug that is specific to locations in additional documents, not source documents: https://github.com/dotnet/roslyn-analyzers/blob/5102a0d89ef178b0ca741634ee006597bf5df2ed/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs#L533-L535

I agree with your prior suggestion though, I think all the fix related locations can be computed directly in the code fixer. However, if you do not want to change the code significantly and want to still pass in 2 additional locations for code fixer, you should directly pass in the locations into the `additionalLocations` parameter taken by Diagnostic.Create method instead of using properties to pass this data: https://github.com/dotnet/roslyn-analyzers/blob/5102a0d89ef178b0ca741634ee006597bf5df2ed/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs#L550